anasymod/templates/xsct_program.py

```diff
@@ -38,40 +87,80 @@ def line(self, s='', nl='\n'):
     def puts(self, s):
         self.line(f'puts "{s}"')
 
+    def set_target(self, pattern):
+        self.line(f'targets -set -filter {{name =~ "{pattern}"}}')
+
     def connect(self):
         self.puts('Connecting to the HW server...')
         self.line('connect')
         self.line()
 
-    def select_cpu(self, cpu_filter):
-        self.puts('Selecting the CPU...')
-        self.line(f'targets -set -filter {{name =~ {cpu_filter}}}')
-        self.line()
-
     def reset_system(self):
         self.puts('Resetting the system...')
-        self.line('rst')
+        self.set_target('APU*')
+        if self.is_ultrascale:
+            self.line('rst -system')
+            self.line('after 3000')
+        else:
+            self.line('stop')
+            self.line('rst')
         self.line()
 
-    def program_fpga(self, bit_path):
+    def program_fpga(self):
         self.puts('Programming the FPGA...')
-        self.line(f'fpga "{bit_path}"')
+        if self.is_ultrascale:
+            self.set_target('PSU')
+        else:
+            self.set_target('xc7z*')
+        self.line(f'fpga "{self.bit_path}"')
         self.line()
 
     def loadhw(self, hw_path):
         self.puts('Setting up the debugger...')
-        self.line(f'loadhw "{hw_path}"')
+        self.set_target('APU*')
+        cmd = []
+        cmd += ['loadhw']
+        cmd += ['-hw', f'"{hw_path}"']
+        if self.is_ultrascale:
+            cmd += ['-mem-ranges [list {0x80000000 0xbfffffff} '
```

Comment:
these ranges do not appear to be program-dependent, since they appear in the global file ``zynqmp_utils.tcl``.  In fact, I tried removing the ``mem-ranges`` argument altogether, and it still worked, but decided to keep this because it replicates the arguments used by the Vitis GUI.